### ISSUE TITLE
ActorMaterializerSettings can't be overriden easily with Guice

### ISSUE THREAD

**Issue by schmitch** [playframework/playframework#6872]
Actually currently we automatically create a `Materializer` that is injectable.
However a Supervision strategy for `Materializer` can't be set with a configuration variable.
So somebody would actually need a way set a `ActorMaterializerSettings` manually.

The only way to do it know is actually by creating a `GuiceApplicationLoader` and override the Settigns via:

```scala

@Singleton
class MaterializerProvider @Inject()(system: ActorSystem) extends Provider[Materializer] {

  private val logger = Logger("play.akka")

  private val decider: Supervision.Decider = { e =>
    logger.error("Unhandled exception in stream", e)
    Supervision.Stop
  }

  private val settings = ActorMaterializerSettings(system).withSupervisionStrategy(decider)

  override lazy val get: Materializer = ActorMaterializer(settings)(system)

}

new GuiceApplicationLoader {
  private val materializerOverrides: Seq[GuiceableModule] = Seq(
    bind[Materializer].toProvider[MaterializerProvider]
  )

  protected override def overrides(context: ApplicationLoader.Context): Seq[GuiceableModule] = {
    GuiceApplicationLoader.defaultOverrides(context) ++ materializerOverrides
  }
}
```

So actually `withSupervisionStrategy` is the only setting that can't be overwritten with a configuration.
We now add either a Module that can be disabled which only contains a `Provider` for `ActorMaterializerSettings`  or we add a Configuration variable to set the SupervisionStrategy, could be done like a Application Loader:
```scala
trait PlaySupervisionStrategy {
   def decider: SupervisionDecider
}
```
and then
`play.supervision.strategy = MySupervisionStrategy` and default to `Supervision.stoppingDecider`

Changing that is super useful for logging errors that occured in streams. So maybe we actually even create a custom Play Supervision by default which logs errors like that:

http://stackoverflow.com/a/35632148/2250209

**Comment by gmethvin** [comment]
@schmitch Could we just create a binding for `ActorMaterializerSettings` and inject it into the `MaterializerProvider`? Seems simpler than creating another trait.

**Comment by schmitch** [comment]
we can't just create a binding for `ActorMaterializerSettings` we would need to create a `Module` as well since without the `Module` we would end up like before, creating a custom `ApplicationLoader`.

**Comment by gmethvin** [comment]
Yeah you're right. I forgot it's not that simple to change bindings with Play.

**Comment by gmethvin** [comment]
We could allow overriding the `ActorMaterializerProvider` through configuration, then have a `DefaultMaterializerProvider` with protected methods you can override. That way you could change anything you want in the `ActorMaterializerSettings`.

**Comment by schmitch** [comment]
Hm... Yeah I think I can do it like the `HttpErrorHandler`: https://github.com/playframework/playframework/blob/26b57b932c617ddb135de202329fc6d8b49ed4be/framework/src/play/src/main/scala/play/api/http/HttpErrorHandler.scala#L48
or the other `dynamicBindings`.
Guess that makes the most sense.
